### ISSUE TITLE
(addon-controls) Support destructering object properties into individual controls

### ISSUE THREAD

**Issue by pimlie** [storybookjs/storybook#12078]
**Is your feature request related to a problem? Please describe.**

Im often passing objects to my (Vue) components which by default will be added as JSON.stringified text control by `@storybook/addon-controls`. As I feel this is confusing/limited functionality (also because the JSON object isnt pretty printed (ie `JSON.stringify(obj, null, 2)`) I'd like to destructure my object properties into separate controls for each individual object property.

I have currently implemented this as follows in this stripped-down User example below
<details><summary>User.vue</summary>

```js
// User.vue
<template>
  <p>The name of the user is: {{ user.name }}</p>
</template>

<script>
export default {
  props: {
    user: {
      type: Object,
      required: true
    }
  }
}
</script>
```
</details>

<details><summary>User.story.js</summary>

```js
import User from '../User.vue'

export default {
  title: 'User',
  component: User,
  argTypes: {
    user_name: {
      control: 'text',
      defaultValue: 'pimlie',
      table: {
        category: 'user',
        type: {
          summary: 'text',
        }
      }
    }
  }
}

const Template = (args, { argTypes }) => {
  console.log(argTypes)
  return {
    components: {
      User
    },
    props: Object.keys(argTypes),
    watch: {
      user_name: {
        immediate: true,
        handler(value) {
          // use $set to ensure reactivity in case name
          // prop doesnt exist on this.user yet
          this.$set(this.user, 'name', value)
        }
      }
    },
    template: '<User :user="user" />',
  }
}

export const Primary = Template.bind({});
Primary.args = {
  user: {
    name: ''
  }
}

```
</details>

(Note: in reality I have abstracted creating the argTypes & watchers to also easily support nested objects, eg to support a dot-path like `blog.message.user.name`)

**Describe the solution you'd like**
I would like that `addon-controls` would have improved support for the above (or a similar) strategy as the current approach of stringifying any object prop doesnt seem very user friendly

Also the above approach has several disadvantages:
- in your _ArgsTable_ you will have both a  `user` control with type object as a `user_name` control with type text
- the actual values of these 2 controls are in sync, so if you change the value of `user_name` the rendered component will show the new value. **But** the stringified value in the `user` object control will not be in sync, it will only show the second to last update. To explain by example:
  - Start with `user` & `user_name` both with value `pimlie` for name
  - Change `user_name` control to `pimlie2`
  - The rendered component updates to `pimlie2`, but the `user` object control will still show `pimlie`
  - Change `user_name` control to `pimlie` again
  - The rendered component updates to `pimlie`, but the `user` object control will now show `pimlie2`
- atm it seems you are unable to change the name of a custom `argType` in the table. See screenshot below, you can group the `user_name` prop but I would also like the label in the `ArgTable` to read just _name_ instead of _user\_name_

**Describe alternatives you've considered**
see above

**Are you able to assist bring the feature to reality?**
sure, I can help with testing

**Additional context**
![image](https://user-images.githubusercontent.com/1067403/90391393-5dfb9200-e08d-11ea-9029-f0e07ca82909.png)

**Comment by Gieted** [comment]
@shilman So what you think about my solution described in #12362? I can maybe PR that, if you accept my specs

**Comment by shilman** [comment]
@Gieted I like the proposal, but I need a few days to let it roll around in my head. Thanks for thinking the problem through and being patient about this!

**Comment by shilman** [comment]
What do you think about this as a workaround: https://github.com/storybookjs/storybook/pull/12685

**Comment by shilman** [comment]
By default, each field in a component's args is rendered as a single row. If a field is a complex type, it will get rendered using a JSON editor per #12599 (existing behavior but better UX).

As an opt-in, the user can also specify a custom argType:

```js
export default {
  title: ...,
  component: ...,
  argTypes: {
    foo: { table: { expanded: true } }
  }
}
```

In this case, the `ArgsTable` will generate an extra row for each field of `foo` with design TBD.

**Comment by vladimircores** [comment]
I duplicate our use-case from https://github.com/storybookjs/storybook/issues/12362

This is exactly what we need in our case where we feed our components with VOs wrapper around real data (JSON) with additional properties from other parts of the system, or precomputed (so component won't know about the store or any data transformation logic):

So VO looks like this:
```
class OrganizationInfoVO {
  constructor(data, isDefaultOrganization, isCurrentUserOrganization) {
    this.data = data;
    this.isDefaultOrganization = isDefaultOrganization;
    this.isCurrentUserOrganization = isCurrentUserOrganization;
  }

  get id() { return this.data.id; }

  get name() { return this.data.name; }

  get country() { return this.data.country; }

  get picture() { return this.data.picture; }

  get count() { return this.data.count; }
}

export default OrganizationInfoVO;
```
And it would be great to have a way of changing properties or visualize them in storybook panels.
I think having at least **some kind of "update callback" from control which called before component re-render** would be great.

**Comment by jpmarra** [comment]
> By default, each field in a component's args is rendered as a single row. If a field is a complex type, it will get rendered using a JSON editor per #12599 (existing behavior but better UX).
> 
> As an opt-in, the user can also specify a custom argType:
> 
> ```js
> export default {
>   title: ...,
>   component: ...,
>   argTypes: {
>     foo: { table: { expanded: true } }
>   }
> }
> ```
> 
> In this case, the `ArgsTable` will generate an extra row for each field of `foo` with design TBD.

I'm not seeing this functionality. I have the custom `argTypes` configuration as listed above, but it is still showing the JSON editor instead of a row for each field.

**Comment by shilman** [comment]
@jpmarra the issue is still open, meaning that it hasn't been added to storybook yet

**Comment by jpmarra** [comment]
@shilman Ah I misread, my bad!

**Comment by iamtomhanks** [comment]
Also really wanting this feature for a client. It seems like an obvious one. Passing a large object as a prop happens quite often. Or arrays of objects as well. That would another good one.

1. **An object prop** - A control for editing properties of objects.
2. **An array of objects prop** A control for editing adding an object to the array and then editing properties of the objects.

**Comment by isc30** [comment]
please, this is such a great improvement

**Comment by annacv** [comment]
🙏

**Comment by vladimircores** [comment]
It would be good if restructuring will be happening from TypeScript annotations.

**Comment by samuelpietra** [comment]
Waiting on it

**Comment by shilman** [comment]
We'll be iterating this PR and releasing in 6.2 as a short-term improvement: https://github.com/storybookjs/storybook/pull/12824

**Comment by gpessa** [comment]
The solution doesn't work with autogenerated controls.
Would be nice to have something native, coming from storybook

**Comment by AthiNos** [comment]
@ashnamuh thanks for the suggestion Flattening and unflattening worked wonders for my nested objects ❤️

**Comment by AmirTugi** [comment]
Is there any progress with this?

**Comment by BrendanRomanDev** [comment]
This seems so essential! I was shocked when reading the docs that this is not available :(

**Comment by tinesoft** [comment]
Hi. Any update on this  @shilman ? It would be very valuable indeed

**Comment by Victor-Nyagudi** [comment]
I'm facing the same problem. 

Unlike most people, my aim is to specify a nested property as an `action` in Storybook. 

Either way, I haven't found anything in the docs or heard about a fix for this coming soon.

**Comment by javimsevilla** [comment]
> [...] my aim is to specify a nested property as an `action` in Storybook. [...]

@Victor-Nyagudi, I faced the same problem recently and found a way to solve it. It's not an automatic way to fix it, but it worked for me.

First, import the action factory from the Storybook action library:

```js
import { action } from '@storybook/addon-actions'
```

Then, in the arguments of your story, create an action by calling the factory method, passing it the name you want to be displayed in Storybook when calling the action:

```js
const onChangeAction = action('onChange') // the name 'onChange' can be anything you want
```

Finally, run the function you created earlier, passing it the values you want to display in Storybook:

```js
// const value = "foo"
onChangeAction({ value }) // this will display 'onChange { value: "foo" }' in Storybook
```

Example:

```js
import { action } from '@storybook/addon-actions'

// [...]

const SomeExampleStory = // ...

SomeExampleStory.args = {
  someDeepObjectProp: {
    onChange({ target: { value } }) {
      const onChangeAction = action('onChange')
      onChangeAction({ value })
    }
  },
  // ...other props...
}
```

Of course, you need `@storybook/addon-actions` to be installed as a dependency in your project, or [@storybook/addon-essentials](https://www.npmjs.com/package/@storybook/addon-essentials) (because `@storybook/addon-actions` is part of it).

> ⚠️ **IMPORTANT**: Please do not install `@storybook/addon-actions` *and* `@storybook/addon-essentials` at the same time or it will not work.

**Comment by sw-tracker** [comment]
+1 for this feature

**Comment by gpessa** [comment]
+1

**Comment by younessadmi** [comment]
+1

**Comment by Tech-Code1** [comment]
+1

**Comment by jen-n-taylor** [comment]
+100000

**Comment by RemLawrence** [comment]
Wow, I didn't realize this feature is not supported at all! What a bummer.

**Comment by ssblake** [comment]
I just ran into the need for this feature as well...

**Comment by frankwaalkens** [comment]
Me too, often I pass a single object as a prop to a component, instead of passing all props seperate: 

`data={dog}`
instead of
`name={dog.name} breed={dog.breed} ...etc`

**Comment by eliasm307** [comment]
Hi guys, I have also been frustrated with this and I've made an addon to make things a bit easier https://www.npmjs.com/package/storybook-addon-deep-controls , feedback welcome

**Comment by omarkawach** [comment]
+1

**Comment by timsofteng** [comment]
+1

**Comment by fastndead** [comment]
+1

**Comment by ElRaymond0** [comment]
There seems to be a lot of support for this feature, do we have an update on whether this is being looked at or considered as a feature?

**Comment by AnandaArya** [comment]
+ 1

**Comment by LeakStephane** [comment]
+1

**Comment by rmmssn** [comment]
+1

**Comment by eric-gitta-moore** [comment]
+1

**Comment by danielzlo** [comment]
+1

**Comment by alexis89x** [comment]
+1

**Comment by beltechsoft** [comment]
+1

**Comment by basit-qayoom** [comment]
+1000

**Comment by AinaVendrell** [comment]
+1

**Comment by AnandaArya** [comment]
still waiting for this one

**Comment by p-hubbell** [comment]
+1

**Comment by IchabodCrow** [comment]
+1

**Comment by johndetlefs-uts** [comment]
+1

**Comment by dwong-conf** [comment]
+1

**Comment by urrri-redis** [comment]
+1

**Comment by eliasm307** [comment]
Just a reminder there is a package that provides this functionality, before it is natively supported by Storybook https://www.npmjs.com/package/storybook-addon-deep-controls

**Comment by viccaruso** [comment]
+1

**Comment by elafontaine** [comment]
+1, Although I'm interested into getting this supported for injecting an array of object having a certain structure.
